### ISSUE TITLE
ember/no-side-effects should allow modifying locale variables

### ISSUE THREAD

**Issue by mydea** [ember-cli/eslint-plugin-ember#377]
E.g. the following code:

```js
options: computed(function() {
  let options = {};
  this._baseOptions.forEach((opt) => {
    set(options, opt, true);
  });

  return options;
})
```

should IMHO not trigger the `no-side-effects` rule. I'm not quite sure if/how this can be realized. But I guess it could at least work for local variables that do not include `this`? E.g.

```js
// allowed
let options = {};

// not allowed
let options = this.baseOptions;
let options = get(this, 'nested.thing');
```

An alternate way could be to just disallow `set(this, ...)`? So e.g. `set(otherThing, ...)` could be allowed. Maybe we could allow to specify an optional list of properties not allowed to set on, for example:

```
"ember/no-side-effects": {
  "disallowOn": ['this']
}
```

**Comment by Turbo87** [comment]
is there a reason why you use `set` instead of `options[opt] = true`?

**Comment by mydea** [comment]
That example was a bit simplified, in reality this is an EmberObject!

**Comment by rwjblue** [comment]
I don't know that it is possible for us to tell the difference between:

```js
computed(function() {
  let { thing } = this;

  set(thing, 'foo', 'bar');
})
```

and 

```js
computed(function() {
  let thing = somefunc();
  set(thing, 'foo', 'bar');
}
```

**Comment by bmish** [comment]
Related issue: #351
